shell/server/mail.js

```diff
@@ -535,6 +538,256 @@ HackSessionContextImpl.prototype.revokeApiToken = function (tokenId) {
   }).bind(this));
 }
 
+function IpInterfaceImpl () { }
+
+IpInterfaceImpl.prototype.listenTcp = function (portNum, port) {
+  return new Promise(function (resolve, reject) {
+    var resolved = false;
+    var server = Net.createServer(function (connection) {
+      connection.done = connection.end;
+      var upstream = port.connect(connection).upstream;
+
+      connection.on('data', function (data) {
+        upstream.write(data);
+      });
+
+      connection.on('close', function (had_error) {
+        upstream.done();
+      });
+
+      connection.on('error', function (err) {
+        if (resolved) {
+          connection.write = errorWrite;
+        } else {
+          // upstream hasn't been resolved yet, so it's safe to reject
+          reject(err);
+        }
+      });
+    });
+
+    server.listen(portNum, function () {
+      resolved = true;
+      resolve({handle: server}); // server has a close method which is all we want from a handle
+    });
+
+    server.on('error', function (err) {
+      if (!resolved) {
+        reject(err);
+      }
+    });
+  });
+
+};
+
+function BoundUdpPortImpl(server, address, port) {
+  this.server = server;
+  this.address = address;
+  this.port = port;
+}
+
+BoundUdpPortImpl.prototype.send = function(message, returnPort) {
```

Comment:
Comment that we're assuming that `returnPort` is always the one corresponding to `this.server`, but that we should really be maintaining a map, but that that requires capability equality which we don't have yet.